mhkit/river/io/d3d.py

```diff
@@ -0,0 +1,444 @@
+from os.path import abspath, dirname, join, isfile, normpath, relpath
+from math import isclose
+import scipy.interpolate as interp
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import netCDF4
+
+
+def get_layer_data(data, variable, layer_index= -1 , time_index=-1):
+    '''
+    Get variable data from netcdf4 object at specified layer and timestep. 
+
+    Parameters
+    ----------
+    data: netcdf4 object 
+       A netCDF object that contains spatial data, e.g. velocity or shear stess.
+       These can be generated by running a Delft3D model.  
+    variable: string
+        Delft3D outputs many vairables that can be called. The full list can be
+        found using "data.variables.keys()" in the consol. 
+    layer_index: int
+         A positve interger to pull out a layer from the dataset. 0 being closest 
+         to the surface. Defalt is the bottom layer -1.
+    time_index: int 
+        A positive interger to pull the time step from the dataset. 0 being closest
+        to time 0. Defalt is last time step -1.  
+
+    Returns
+    -------
+    layer_data: DataFrame
+        DataFrame with columns of "x" and "y" location on specified layer and 
+        the variable values "v".
+    '''
+    
+    assert isinstance(time_index, int), 'time_index  must be a int'
+    assert isinstance(layer_index, int), 'layer_index  must be a int'
+    assert type(data)== netCDF4._netCDF4.Dataset, 'data must be netCDF4 object'
+    assert variable in data.variables.keys(), 'variable not recognized'
+    coords = str(data.variables[variable].coordinates).split()
+    var=data.variables[variable][:]
+    max_time_index= len(var)
+    assert time_index <= max_time_index, (f'time_index must be less than the max'
+                                         +'time index {max_time_index}')
+    assert time_index >= -1, 'time_index must be greater than or equal to -1'
+    max_layer= len(var[0][0])
+    assert layer_index <= max_layer,( f'layer_index must be less than the max'
+                                     + 'layer {max_layer}')
+    assert layer_index >= -1, 'layer_index must be greater than or equal to -1'
+    
+    x=np.ma.getdata(data.variables[coords[0]][:], False) 
+    y=np.ma.getdata(data.variables[coords[1]][:], False)
+ 
+    v= np.ma.getdata(var[time_index,:,layer_index], False)
+    
+    layer= np.array([ [x_i, y_i, z_i] for x_i, y_i, z_i in zip(x, y, v)]) 
+    layer_data = pd.DataFrame(layer, columns=[ 'x', 'y', 'v'])
+    
+    return layer_data
+
+
+def create_points(x, y, z):
+    '''
+    Turns x, y, and z into a DataFrame of points to interpolate over.
```

Comment:
This could be a bit more descriptive to make the use case more clear. Might just be as simple as pointing users to the example notebook for example of use.